File: java/datamover/src/main/java/ai/tonic/datamover/DbConfig.java
package ai.tonic.datamover;

import org.tomlj.TomlTable;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Objects;
import java.util.Properties;

public record DbConfig(String host, int port, String user, String password, String database)
{
    public static DbConfig readConfigFrom(TomlTable config, String type)
    {
        var databaseConfig = config.getTable(type);
        var host = databaseConfig.getString("host");
        var port = Objects.requireNonNull(databaseConfig.getLong("port")).intValue();
        var user = databaseConfig.getString("user");
        var password = databaseConfig.getString("password");
        var database = databaseConfig.getString("database");
        return new DbConfig(host, port, user, password, database);
    }

    public Connection connection()
    {
        String url = "jdbc:postgresql://" + host + ":" + port + "/" + database;
        Properties props = new Properties();
        props.setProperty("user", user);
        props.setProperty("password", password);
        try {
            return DriverManager.getConnection(url, props);
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
    }
}

File: dotnet/data_mover/DbConfig.cs
using Npgsql;
using Tomlyn.Model;

namespace data_mover;

public record DbConfig(string Host, int Port, string User, string Password, string Database)
{
    public static DbConfig ReadConfigFrom(TomlTable config, string type)
    {
        var databaseConfig = config[type] as TomlTable;
        if (databaseConfig is null)
        {
            throw new ArgumentException("Configuration missing database type: " + type);
        }
        var host = (string) databaseConfig["host"];
        var port = Convert.ToInt32((long) databaseConfig["port"]);
        var user = (string) databaseConfig["user"];
        var password = (string) databaseConfig["password"];
        var database = (string) databaseConfig["database"];

        return new DbConfig(host, port, user, password, database);
    }

    public NpgsqlConnection Connection()
    {
        var stringBuilder = new NpgsqlConnectionStringBuilder
        {
            Database = Database,
            Host = Host,
            Port = Port,
            Username = User,
            Password = Password
        };

        return new NpgsqlConnection(stringBuilder.ToString());
    }
}